lib/fetch/headers.js

```diff
@@ -91,64 +77,68 @@ function fill (headers, object) {
   }
 }
 
-// TODO: Composition over inheritence? Or helper methods?
-class HeadersList extends Array {
+class HeadersList {
+  constructor (init) {
+    if (init instanceof HeadersList) {
+      this[kHeadersMap] = new Map(init[kHeadersMap])
+      this[kSorted] = init[kSorted]
+    } else {
+      this[kHeadersMap] = new Map(init)
+      this[kSorted] = false
+    }
+  }
+
   append (name, value) {
     const normalizedName = normalizeAndValidateHeaderName(name)
     const normalizedValue = normalizeAndValidateHeaderValue(name, value)
 
-    const index = binarySearch(this, normalizedName)
+    const exists = this[kHeadersMap].get(normalizedName)
 
-    if (this[index] === normalizedName) {
-      this[index + 1] += `, ${normalizedValue}`
+    if (exists) {
+      this[kHeadersMap].set(normalizedName, `${exists}, ${normalizedValue}`)
     } else {
-      this.splice(index, 0, normalizedName, normalizedValue)
+      this[kHeadersMap].set(normalizedName, `${normalizedValue}`)
+      this[kSorted] = false
     }
   }
 
   delete (name) {
     const normalizedName = normalizeAndValidateHeaderName(name)
-
-    const index = binarySearch(this, normalizedName)
-
-    if (this[index] === normalizedName) {
-      this.splice(index, 2)
-    }
+    return this[kHeadersMap].delete(normalizedName)
   }
 
   get (name) {
     const normalizedName = normalizeAndValidateHeaderName(name)
-
-    const index = binarySearch(this, normalizedName)
-
-    if (this[index] === normalizedName) {
-      return this[index + 1]
-    }
-
-    return null
+    return this[kHeadersMap].get(normalizedName) ?? null
   }
 
   has (name) {
     const normalizedName = normalizeAndValidateHeaderName(name)
+    return this[kHeadersMap].has(normalizedName)
+  }
 
-    const index = binarySearch(this, normalizedName)
+  sortMap () {
+    this[kHeadersMap] = new Map([...this[kHeadersMap]].sort((a, b) => a[0] < b[0] ? -1 : 1))
+    this[kSorted] = true
+  }
 
-    return this[index] === normalizedName
+  // bypass normalization
+  set (key, val) {
```

Comment:
Why are we bypassing normalizations?

Reply:
It's a way to improve the performance once it wouldn't pass by validations. I did it just in cases where we assure the key/value are valid already

Reply:
Can you make a safe and unsafeSet? To make it a bit more explicit.

Reply:
I'd prefer if you keep `append` where used so that it matches the spec text.

Reply:
That's fair. This micro-optimization shouldn't affect too much. I'll adjust to using `append` everywhere.